### PR TITLE
DRILL-3871: Off by one error while reading binary fields with one ter…

### PR DESCRIPTION
**Pull request by parthchandra**
…minal null in parquet.

Changes -
  1) Rewrote the NullableColumnReader.processPages function to process runs of Null values and Non-Null values without needing to keeping track of whether the previous iteration in the while loop had encountered a null or not. A pair of loops now iterates over a run of nulls or a run of non-null values.
  2) Removed some redundant code.
  3) Renamed some variables. The indexInOutputVector is now replaced by two local variables, readCount and writeCount only for clarity.
 4) Adding tracing.
 5) Added unit tests for edge cases of nulls occurring on page boundaries. 

For all the unit tests, tpch-h and tpc-ds test data sets, the state of the NullableColumnReader at the end of each iteration of processPages is identical to the old code. In addition the boundary conditions are taken care of.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java [hunk 1]
```diff
@@ -29,117 +29,152 @@
...
         }
-        valuesReadInCurrentPass += nullsFound;
+      }
+      //
+      // Write the nulls if any
+      //
+      if (nullRunLength > 0) {
+        int writerIndex =
+            ((BaseDataValueVector) valueVec).getBuffer().writerIndex();
+        castedBaseVector.getBuffer().setIndex(0, writerIndex + (int) Math
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java]: previous code seemed to handle `dataTypeLengthInBits <= 8` separately. Is it no longer required/needed ?

Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java]: Yes. This code was useful when the Bit data type was a single bit. It is no longer so and this check is not needed.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java [hunk 2]
```diff
@@ -29,117 +29,152 @@
...
+    int currentDefinitionLevel = -1;
+    int readCount = 0; // the record number we last read.
+    int writeCount = 0; // the record number we last wrote to the value vector.
+                        // This was previously the indexInOutputVector variable
+    boolean haveMoreData; // true if we have more data and have not filled the vector
+
+    while (readCount < recordsToReadInThisPass && writeCount < valueVec.getValueCapacity()) {
+      // read a page if needed
       if (!pageReader.hasPage()
-            || ((readStartInBytes + readLength >= pageReader.byteLength && bitsUsed == 0) &&
```
Comment (adeneche) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java]: I'm not seeing this condition in the new code: `readStartInBytes + readhLength >= pageReader.byteLength`

Comment (parthchandra) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/columnreaders/NullableColumnReader.java]: The condition  
readStartInBytes + readLength >= pageReader.byteLength 
and the condition
definitionLevelsRead >= pageReader.currentPageCount
both were checking for the end of a page. I'm not sure if in the previous code one condition could be true but not the other. In the new while loop, if one is true, so is the other. So we need only one of the two.

**Comment by adeneche** [comment]
+1, LGTM

**Comment by jaltekruse** [comment]
+1, thanks for cleaning this up Parth!
